### PR TITLE
Fix _merge_columns when using multiple sensitive features with long names

### PR DESCRIPTION
**Pull request by romanlutz**
Fairlearn supports scenarios with multiple sensitive features (e.g., gender and age, but it can be more than two). Internally, the mitigation techniques currently treat this no different than a single sensitive feature, so there's a column merge step called `_merge_columns()` that happens before applying mitigations.
 
During that step, `numpy.apply_along_axis` has so far been used to merge the string representations of the sensitive features. However, this doesn't return the expected results when the names are long since it cuts them off. For example, @LeJit discovered in #728 that the results from `ThresholdOptimizer` with two sensitive features did not match the results when manually combining the two columns. Since we don't want to restrict users in sensitive feature naming this is not acceptable.

The solution in this PR is to use `pandas.DataFrame.apply` instead of `numpy.apply_along_axis` since it doesn't exhibit the same restrictions. To prevent us from hitting the same issue in the future I've also added a test case which the previous implementation would have failed.

I also suggest we release v0.6.2 once this is in since it might actually affect users (in addition to @LeJit ).

Huge kudos to @LeJit for the excellent repro in #728 . It is a rare pleasure for an engineer to get a simple repro including the specified expectations. :-)

### CODE REVIEW

File: test/unit/metrics/test_metricframe_smoke.py [hunk 1]
```diff
@@ -270,7 +269,6 @@ def multi_sp(y_t, y_p, p1, p2):
 
 def test_2m_1sf_sample_weights():
     """Check that sample weights are passed correctly to two metrics."""
-
```
Comment (riedgar-ms) [inline test/unit/metrics/test_metricframe_smoke.py:273]: ?

File: test/unit/postprocessing/test_threshold_optimizer_multiple_sensitive_features.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    # The bug caused the merged names to get cut off, resulting in multiple groups
+    # getting merged internally. To avoid that this test case checks even internal
+    # representations.
+    X = pd.DataFrame([
+        [0, 4], [6, 2], [1, 3], [10, 5], [1, 7], [-2, 1], [3, 10], [14, 5],
+        [1, 3], [1, 5], [1, 7], [-5, 9], [3, 13], [7, 1], [-8, 4], [9, 1]])
+    y = pd.Series([0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0])
+    a1 = "a"
+    a2 = "a very very very very very very very long group name"
+    a3 = "a group name with commas ,, in , it"
```
Comment (riedgar-ms) [inline test/unit/postprocessing/test_threshold_optimizer_multiple_sensitive_features.py:39]: Would it be worth having one with a trailing backslash too?

**Review by riedgar-ms — CHANGES_REQUESTED** [review]

File: test/unit/metrics/test_metricframe_smoke.py (hunk 1 above)
Comment (romanlutz) [reply test/unit/metrics/test_metricframe_smoke.py:273]: linting issue, now resolved

File: test/unit/postprocessing/test_threshold_optimizer_multiple_sensitive_features.py (hunk 2 above)
Comment (romanlutz) [reply test/unit/postprocessing/test_threshold_optimizer_multiple_sensitive_features.py:39]: A single one will escape the following characters. What do you want to escape?

Comment (riedgar-ms) [reply test/unit/postprocessing/test_threshold_optimizer_multiple_sensitive_features.py:39]: I meant have the embedded string have the trailing backslash (i.e. put in two), just to make sure that nothing then subsequently used that backslash to escape something.

Comment (romanlutz) [reply test/unit/postprocessing/test_threshold_optimizer_multiple_sensitive_features.py:39]: I added backslashes to the test case. I'm not entirely sure whether you meant something in addition to that. If so, can you provide an example?

Comment (riedgar-ms) [reply test/unit/postprocessing/test_threshold_optimizer_multiple_sensitive_features.py:39]: Looks good

**Review by riedgar-ms — APPROVED** [review]

**Review by MiroDudik — APPROVED** [review]

**Pull request merged**